File: javascript/packages/core/src/policy.js
/** Authorisation rules. */

import { RepairStatus } from "./support/status.js";

/**
 * Who is asking.
 *
 * @typedef {{ role: "customer" | "technician" | "manager", id?: number }} Actor
 */

/** Gate around repair-order mutations. */
export class RepairOrderPolicy {
    /**
     * Customers see only their own orders; staff see everything.
     *
     * @param {Actor} actor
     * @param {import("./models/repairOrder.js").RepairOrder} order
     * @returns {boolean}
     */
    canView(actor, order) {
        return actor.role === "customer" ? actor.id === order.customerId : true;
    }

    /**
     * Only staff move an order forward, and never past a terminal state.
     *
     * @param {Actor} actor
     * @param {import("./models/repairOrder.js").RepairOrder} order
     * @returns {boolean}
     */
    canTransition(actor, order) {
        if (RepairStatus.isTerminal(order.status)) {
            return false;
        }

        return actor.role !== "customer";
    }

    /**
     * Collection is signed off by the owning customer or a manager.
     *
     * @param {Actor} actor
     * @param {import("./models/repairOrder.js").RepairOrder} order
     * @returns {boolean}
     */
    canCollect(actor, order) {
        if (order.status !== RepairStatus.Completed) {
            return false;
        }

        switch (actor.role) {
            case "customer":
                return actor.id === order.customerId;
            case "manager":
                return true;
            default:
                return false;
        }
    }
}

File: rust/crates/atelier-core/src/policy.rs
//! Authorisation rules.

use crate::models::repair_order::RepairOrder;
use crate::support::status::RepairStatus;

/// Who is asking.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Actor {
    Customer { id: u32 },
    Technician { id: u32 },
    Manager,
}

/// Gate around [`RepairOrder`] mutations.
#[derive(Debug, Clone, Copy, Default)]
pub struct RepairOrderPolicy;

impl RepairOrderPolicy {
    /// Customers see only their own orders; staff see everything.
    pub fn can_view(&self, actor: Actor, order: &RepairOrder) -> bool {
        match actor {
            Actor::Customer { id } => id == order.customer_id,
            Actor::Technician { .. } | Actor::Manager => true,
        }
    }

    /// Only staff move an order forward, and never past a terminal state.
    pub fn can_transition(&self, actor: Actor, order: &RepairOrder) -> bool {
        if order.status.is_terminal() {
            return false;
        }

        matches!(actor, Actor::Technician { .. } | Actor::Manager)
    }

    /// Collection is signed off by the owning customer or a manager.
    pub fn can_collect(&self, actor: Actor, order: &RepairOrder) -> bool {
        if order.status != RepairStatus::Completed {
            return false;
        }

        match actor {
            Actor::Customer { id } => id == order.customer_id,
            Actor::Manager => true,
            Actor::Technician { .. } => false,
        }
    }
}
